Add ViTPose (#30530)
huggingface/transformers: docs/source/en/model_doc/vitpose.md

This model was contributed by [nielsr](https://huggingface.co/nielsr) and [sangbumchoi](https://github.com/SangbumChoi).
The original code can be found [here](https://github.com/ViTAE-Transformer/ViTPose).

## Usage Tips

- To enable MoE (Mixture of Experts) function in the backbone, user has to give appropriate configuration such as `num_experts` and input value `dataset_index` to the backbone model. 
  However, it is not used in default parameters. Below is the code snippet for usage of MoE function.
```py
>>> from transformers import VitPoseBackboneConfig, VitPoseBackbone
>>> import torch

>>> config = VitPoseBackboneConfig(num_experts=3, out_indices=[-1])
>>> model = VitPoseBackbone(config)

>>> pixel_values = torch.randn(3, 3, 256, 192)
>>> dataset_index = torch.tensor([1, 2, 3])
>>> outputs = model(pixel_values, dataset_index)
```

- ViTPose is a so-called top-down keypoint detection model. This means that one first uses an object detector, like [RT-DETR](rt_detr.md), to detect people (or other instances) in an image. Next, ViTPose takes the cropped images as input and predicts the keypoints.

```py

> import math
> 
> import cv2

Reviewer: ```suggestion
```

No longer needed, right?

> import cv2

Reviewer: (when copying the code for the first block into my notebook, it threw an error saving cv2 is not installed, even though I was only wanting to run the section without the dependency)

> import cv2

Reviewer 2: done!

    to (top_left_x, top_left_y, width, height).

    Args:
        bboxes (`np.ndarray` of shape `(batch_size, 4)):
            Bounding boxes in Pascal VOC format.

    Returns:
        `np.ndarray` of shape `(batch_size, 4) in COCO format.
    """
    bboxes[:, 2] = bboxes[:, 2] - bboxes[:, 0]
    bboxes[:, 3] = bboxes[:, 3] - bboxes[:, 1]

    return bboxes

# 0 index indicates human label in COCO
boxes = results[0]['boxes'][results[0]['labels'] == 0]
boxes = [pascal_voc_to_coco(boxes.cpu().numpy())]

image_processor = VitPoseImageProcessor.from_pretrained("nielsr/vitpose-base-simple")
model = VitPoseForPoseEstimation.from_pretrained("nielsr/vitpose-base-simple")

# Stage 2. Run ViTPose
pixel_values = image_processor(image, boxes=boxes, return_tensors="pt").pixel_values

with torch.no_grad():
  outputs = model(pixel_values)

pose_results = image_processor.post_process_pose_estimation(outputs, boxes=boxes)[0]

for pose_result in pose_results:
  for keypoint in pose_result['keypoints']:
    x, y, score = keypoint
    print(f"coordinate : [{x}, {y}], score : {score}")

def visualize_keypoints(img,

>   pose_result,

Reviewer 2: fixed

  skeleton=None,
  kpt_score_thr=0.3,
  pose_kpt_color=None,
  pose_link_color=None,
  radius=4,
  thickness=1,
  show_keypoint_weight=False):
    """Draw keypoints and links on an image.

>     Args:

Reviewer 2: all updated

            img (str or Tensor): The image to draw poses on. If an image array
                is given, id will be modified in-place.
            pose_result (list[kpts]): The poses to draw. Each element kpts is
                a set of K keypoints as an Kx3 numpy.ndarray, where each
                keypoint is represented as x, y, score.
            kpt_score_thr (float, optional): Minimum score of keypoints
                to be shown. Default: 0.3.
            pose_kpt_color (np.array[Nx3]`): Color of N keypoints. If None,
                the keypoint will not be drawn.
            pose_link_color (np.array[Mx3]): Color of M links. If None, the
                links will not be drawn.

>             thickness (int): Thickness of lines.

Reviewer 2: all updated

    """
    img = img.copy()
    img_h, img_w, _ = img.shape

    for kpts in pose_result:
        kpts = np.array(kpts, copy=False)

>         # draw each point on image

Reviewer 3: let's split up some of this functionality into smaller function i.e. this function calls `draw_points` and `draw_links`

Reviewer 2: no problem

                        cv2.addWeighted(
                            img_copy,
                            transparency,
                            img,
                            1 - transparency,
                            0,
                            dst=img)
                    else:
                        cv2.circle(img, (int(x_coord), int(y_coord)), radius,
                                   color, -1)

        # draw links
        if skeleton is not None and pose_link_color is not None:
            assert len(pose_link_color) == len(skeleton)
            for sk_id, sk in enumerate(skeleton):
                pos1 = (int(kpts[sk[0], 0]), int(kpts[sk[0], 1]))
                pos2 = (int(kpts[sk[1], 0]), int(kpts[sk[1], 1]))
                if (pos1[0] > 0 and pos1[0] < img_w and pos1[1] > 0
                        and pos1[1] < img_h and pos2[0] > 0 and pos2[0] < img_w
                        and pos2[1] > 0 and pos2[1] < img_h
                        and kpts[sk[0], 2] > kpt_score_thr
                        and kpts[sk[1], 2] > kpt_score_thr):
                    color = tuple(int(c) for c in pose_link_color[sk_id])
                    if show_keypoint_weight:

>                         img_copy = img.copy()

Reviewer 2: It is not necessary so I removed all

                            dst=img)
                    else:
                        cv2.line(img, pos1, pos2, color, thickness=thickness)

    return img

# Note: skeleton and color palette are dataset-specific
skeleton = [[15, 13], [13, 11], [16, 14], [14, 12], [11, 12],
                        [5, 11], [6, 12], [5, 6], [5, 7], [6, 8], [7, 9],
                        [8, 10], [1, 2], [0, 1], [0, 2], [1, 3], [2, 4],
                        [3, 5], [4, 6]]

palette = np.array([[255, 128, 0], [255, 153, 51], [255, 178, 102],
                            [230, 230, 0], [255, 153, 255], [153, 204, 255],
                            [255, 102, 255], [255, 51, 255], [102, 178, 255],
                            [51, 153, 255], [255, 153, 153], [255, 102, 102],
                            [255, 51, 51], [153, 255, 153], [102, 255, 102],
                            [51, 255, 51], [0, 255, 0], [0, 0, 255],
                            [255, 0, 0], [255, 255, 255]])

pose_link_color = palette[[
                0, 0, 0, 0, 7, 7, 7, 9, 9, 9, 9, 9, 16, 16, 16, 16, 16, 16, 16
            ]]
pose_kpt_color = palette[[
                16, 16, 16, 16, 16, 9, 9, 9, 9, 9, 9, 0, 0, 0, 0, 0, 0

> ]]

Reviewer 2: fixed by linting

This model was contributed by [nielsr](https://huggingface.co/nielsr) and [sangbumchoi](https://github.com/SangbumChoi).
The original code can be found [here](https://github.com/ViTAE-Transformer/ViTPose).

## Usage Tips

- To enable MoE (Mixture of Experts) function in the backbone, user has to give appropriate configuration such as `num_experts` and input value `dataset_index` to the backbone model. 
  However, it is not used in default parameters. Below is the code snippet for usage of MoE function.
```py
>>> from transformers import VitPoseBackboneConfig, VitPoseBackbone
>>> import torch

>>> config = VitPoseBackboneConfig(num_experts=3, out_indices=[-1])
>>> model = VitPoseBackbone(config)

>>> pixel_values = torch.randn(3, 3, 256, 192)
>>> dataset_index = torch.tensor([1, 2, 3])
>>> outputs = model(pixel_values, dataset_index)
```

> - ViTPose is a so-called top-down keypoint detection model. This means that one first uses an object detector, like [RT-DETR](rt-detr), to detect people (or other instances) in an image. Next, ViTPose takes the cropped images as input and predicts the keypoints.

Reviewer 2: I think it is rather rt_detr.md right?

            Skeleton definition.
        kpt_score_thr (`float`, *optional*, defaults to 0.3): 
            Minimum score of keypoints to be shown.
        pose_kpt_color (`numpy.ndarray`, *optional*): 
            Color of N keypoints. If None, the keypoints will not be drawn.
        pose_link_color (`numpy.ndarray`, *optional*): 
            Color of M links. If None, the links will not be drawn.
        radius (`int`, *optional*, defaults to 4):
            Radius of keypoint circles.
        thickness (`int`, *optional*, defaults to 1): 
            Thickness of lines.
        show_keypoint_weight (`bool`, *optional*, defaults to False): 
            Whether to adjust keypoint and link visibility based on the keypoint scores.
    
    Returns:
        `numpy.ndarray`: Image with drawn keypoints and links.
    """
    for kpts in pose_result:
        kpts = np.array(kpts, copy=False)

        # draw each point on image
        draw_points(pose_kpt_color, kpts, img, kpt_score_thr, radius, show_keypoint_weight)

        # draw links
        draw_links(skeleton, pose_link_color, kpts, img, kpt_score_thr, thickness, show_keypoint_weight)

    return img

> # Note: skeleton and color palette are dataset-specific

Reviewer 2: Yeah that might be more useful. However, this variable is only used in visualizing the output result which is not used in model. Should we still move into model config?/processor?/leave at it is?

> # Note: skeleton and color palette are dataset-specific

Reviewer 4: Good question, I think it's very similar to `id2class` and `class2id` mappings that are stored in config while not used in modeling directly. So, I would argue to add this information as a complementary too.

> # Note: skeleton and color palette are dataset-specific

Reviewer 2: Great will do that :)

<!--Copyright 2024 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# ViTPose

## Overview

> The ViTPose model was proposed in [ViTPose: Simple Vision Transformer Baselines for Human Pose Estimation](https://arxiv.org/abs/2204.12484) by Yufei Xu, Jing Zhang, Qiming Zhang, Dacheng Tao. ViTPose employs a standard [Vision Transformer](vit) as backbone for the task of keypoint estimation.

Reviewer 3: Could you expand this model explanation a bit please?

This model was contributed by [nielsr](https://huggingface.co/nielsr) and [sangbumchoi](https://github.com/SangbumChoi).
The original code can be found [here](https://github.com/ViTAE-Transformer/ViTPose).

## Usage Tips

> - To enable MoE (Mixture of Experts) function in the backbone, user has to give appropriate configuration such as `num_experts` and input value `dataset_index` to the backbone model. 

Author: This part is a bit unclear, this is randomly initializing the model right? What does [1, 2, 3] mean, that all experts will be used?

Not sure anyone will run a randomly initalized ViTPose model

Reviewer 2: @NielsRogge I agree with you. This will not be used unless the finetuning approach is available.

Reviewer 4: relevant for `vitpose-plus-base` checkpoint, which has 6 experts (trained on 6 datasets)

  However, it is not used in default parameters. Below is the code snippet for usage of MoE function.
```py
>>> from transformers import VitPoseBackboneConfig, VitPoseBackbone
>>> import torch

>>> config = VitPoseBackboneConfig(num_experts=3, out_indices=[-1])
>>> model = VitPoseBackbone(config)

>>> pixel_values = torch.randn(3, 3, 256, 192)
>>> dataset_index = torch.tensor([1, 2, 3])
>>> outputs = model(pixel_values, dataset_index)
```

> - The current model utilizes a 2-step inference pipeline. The first step involves placing a bounding box around the region corresponding to the person.
>   After that, the second step uses VitPose to predict the keypoints.

Author: ```suggestion
- ViTPose is a so-called top-down keypoint detection model. This means that one first uses an object detector, like [RT-DETR](rt-detr), to detect people (or other instances) in an image. Next, ViTPose takes the cropped images as input and predicts the keypoints.
```

```py
>>> import torch
>>> import requests

>>> from PIL import Image
>>> from transformers import VitPoseImageProcessor, VitPoseForPoseEstimation

>>> url = 'http://images.cocodataset.org/val2017/000000000139.jpg' 
>>> image = Image.open(requests.get(url, stream=True).raw)

>>> image_processor = VitPoseImageProcessor.from_pretrained("nielsr/vitpose-base-simple")
>>> model = VitPoseForPoseEstimation.from_pretrained("nielsr/vitpose-base-simple")

> >>> boxes = [[[412.8, 157.61, 53.05, 138.01], [384.43, 172.21, 15.12, 35.74]]]

Author: We could add a code snippet which combines RT-DETR with ViTPose

Reviewer 2: Done and added with visualization part!

<!--Copyright 2024 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# ViTPose

## Overview

> The ViTPose model was proposed in [ViTPose: Simple Vision Transformer Baselines for Human Pose Estimation](https://arxiv.org/abs/2204.12484) by Yufei Xu, Jing Zhang, Qiming Zhang, Dacheng Tao. ViTPose employs a standard, non-hierarchical [Vision Transformer](vit) as backbone for the task of keypoint estimation. A simple decoder head is added on top to predict the heatmaps from a given image. Despite its simplicity, the model gets state-of-the-art results on the challenging MS COCO Keypoint Detection benchmark.

Reviewer 2: ```suggestion
The ViTPose model was proposed in [ViTPose: Simple Vision Transformer Baselines for Human Pose Estimation](https://arxiv.org/abs/2204.12484) by Yufei Xu, Jing Zhang, Qiming Zhang, Dacheng Tao. ViTPose employs a standard, non-hierarchical [Vision Transformer](https://arxiv.org/pdf/2010.11929v2) as backbone for the task of keypoint estimation. A simple decoder head is added on top to predict the heatmaps from a given image. Despite its simplicity, the model gets state-of-the-art results on the challenging MS COCO Keypoint Detection benchmark.
```

This model was contributed by [nielsr](https://huggingface.co/nielsr) and [sangbumchoi](https://github.com/SangbumChoi).
The original code can be found [here](https://github.com/ViTAE-Transformer/ViTPose).

## Usage Tips

> - To enable MoE(Mixture or Expert) function in backbone, user have to give appropriate input indices into the backbone model. 

Reviewer 2: @amyeroberts Thanks for pointing out this comment. It was not able to use MoE architecture before. Fixed by
https://github.com/huggingface/transformers/pull/30530/commits/5911010c6e0ac6b6765c521e375278bb683340b4

        x, y, score = keypoint
        print(f"coordinate : [{x}, {y}], score : {score}")

# Visualization for supervision user
import supervision as sv

key_points = sv.KeyPoints(xy=torch.cat([pose_result['keypoints'].unsqueeze(0) for pose_result in pose_results]).cpu().numpy())

edge_annotator = sv.EdgeAnnotator(
    color=sv.Color.GREEN,
    thickness=5
)
annotated_frame = edge_annotator.annotate(
    scene=image.copy(),
    key_points=key_points

> )

Reviewer 2: great idea
